### ISSUE TITLE
`addr x` gives wrong result when `x` is initialized via an if/else expression

### ISSUE THREAD

**Issue by timotheecour** [nim-lang/Nim#11775]
seems like a serious bug; `cast[cstring](addr x)` gives wrong result with float32, when x is initialized via a conditional expression, eg: `var x2 = if x == 1.0: 1.0 else: x`

### Example
nim c -r main.nim # D20190717T192340
```nim
  type Elem = float32

  proc hash3(x: var Elem) =
    var x2 = if x == 1.0: 1.0 else: x
    # var x2 = x replace above line with this and the bug goes away
    var p = cast[cstring](addr x2)
    echo [p[0].int, p[1].int, p[2].int, p[3].int], (x, x2)

  var x = Elem(1.4e-40)
  hash3(x)
  
  x = Elem(12.371)
  hash3(x)
  var x2 = Elem(12.372)
  doAssert x2 != x
  hash3(x2)
```

### Current Output
```
[0, 0, 0, 0](1.399995256752995e-40, 1.399995256752995e-40)
[0, 0, 0, 192](12.37100028991699, 12.37100028991699)
[0, 0, 0, 192](12.37199974060059, 12.37199974060059)
```

### Expected Output
[0,0,0,0] is wrong because the float32 is not 0
the other results are also wrong because [0, 0, 0, 192]==[0, 0, 0, 192] but x != x2

### Additional Information
* Your Nim version (output of `nim -v`).
latest devel at f50e4500c257e3d57654f5230db75042da233d24
* Was it working in the previous Nim releases?
no

**Comment by alaviss** [comment]
Here's a "simplified" test case:

```nim
proc test(x: float32) =
  var a = if x == 1.0: 1.0 else: x

  echo repr(addr a)
  let b = cast[ptr UncheckedArray[byte]](addr a)
  echo repr(b)
  echo b[0], b[1], b[2], b[3]

test(1.0)
var b = 12.371
test(b)
```

The pointer is correct and is the same even via `cast`. Furthermore, the generated C code appears to be correct:

```c
N_LIB_PRIVATE N_NIMCALL(void, test_V4RjF5VLANSEAmxVREFoDw)(NF32 x) {
	NF a;
	tyArray_nHXaesL0DJZHyVS07ARPRA T6_;
	NU8* b;
	tyArray_nHXaesL0DJZHyVS07ARPRA T7_;
	tyArray_sMpvt1sOxOJ3LFGulnbeMQ T8_;
	{
		if (!(((NF) (x)) == 1.0000000000000000e+00)) goto LA3_;
		a = 1.0000000000000000e+00;
	}
	goto LA1_;
	LA3_: ;
	{
		a = ((NF) (x));
	}
	LA1_: ;
	nimZeroMem((void*)T6_, sizeof(tyArray_nHXaesL0DJZHyVS07ARPRA));
	T6_[0] = reprAny((&a), (&NTI_dHulNBWhJTS79cd5h9cp4OHg_));
	echoBinSafe(T6_, 1);
	b = ((NU8*) ((&a)));
	nimZeroMem((void*)T7_, sizeof(tyArray_nHXaesL0DJZHyVS07ARPRA));
	T7_[0] = reprAny(b, (&NTI_al608xe9aGnSEC9culioQEKA_));
	echoBinSafe(T7_, 1);
	nimZeroMem((void*)T8_, sizeof(tyArray_sMpvt1sOxOJ3LFGulnbeMQ));
	T8_[0] = dollar__RkX9btpg5sQIaP8yYXB6tbA(((NU64) (b[((NI) 0)])));
	T8_[1] = dollar__RkX9btpg5sQIaP8yYXB6tbA(((NU64) (b[((NI) 1)])));
	T8_[2] = dollar__RkX9btpg5sQIaP8yYXB6tbA(((NU64) (b[((NI) 2)])));
	T8_[3] = dollar__RkX9btpg5sQIaP8yYXB6tbA(((NU64) (b[((NI) 3)])));
	echoBinSafe(T8_, 4);
}
```

So I'm not sure what's the bug here, but it's definitely not due to `addr`.

**Comment by alaviss** [comment]
Mystery solved: `var a = if x == 1.0: 1.0 else: x` converts x into a float64 because `1.0` is a float constant, which explains the weird output comparing to direct assignment.

Quick fix: `var a = if x == 1.0f32: 1.0f32 else: x`

**Comment by timotheecour** [comment]
wow thanks @alaviss ! a bit of a gotcha, but nothing we can do about it, closing
